File: test/language/array_types/cpp/AutoArrayStructRecursionTest.cpp
#include "gtest/gtest.h"

#include "array_types/auto_array_struct_recursion/AutoArrayRecursion.h"

#include "zserio/SerializeUtil.h"

namespace array_types
{
namespace auto_array_struct_recursion
{

class AutoArrayStructRecursionTest : public ::testing::Test
{
protected:
    void fillAutoArrayRecursion(AutoArrayRecursion& autoArrayRecursion, size_t numElements)
    {
        autoArrayRecursion.setId(0);

        auto& autoArray = autoArrayRecursion.getAutoArrayRecursion();
        for (size_t i = 1; i <= numElements; ++i)
        {
            AutoArrayRecursion element;
            element.setId(static_cast<uint8_t>(i));
            autoArray.push_back(element);
        }
    }

    size_t getAutoArrayRecursionBitSizeOf(size_t numElements)
    {
        return 8 + 8 + numElements * (8 + 8);
    }

    void writeAutoArrayRecursionToByteArray(zserio::BitStreamWriter& writer, size_t numElements)
    {
        writer.writeBits(0, 8);
        writer.writeVarSize(static_cast<uint32_t>(numElements));
        for (size_t i = 1; i <= numElements; ++i)
        {
            writer.writeBits(static_cast<uint32_t>(i), 8);
            writer.writeVarSize(0);
        }
    }

    void checkAutoArrayRecursion(const AutoArrayRecursion& autoArrayRecursion, size_t numElements)
    {
        ASSERT_EQ(0, autoArrayRecursion.getId());
        const auto& autoArray = autoArrayRecursion.getAutoArrayRecursion();
        ASSERT_EQ(numElements, autoArray.size());
        for (size_t i = 1; i <= numElements; ++i)
        {
            const AutoArrayRecursion& element = autoArray.at(i - 1);
            ASSERT_EQ(i, element.getId());
            ASSERT_EQ(0, element.getAutoArrayRecursion().size());
        }
    }

    void checkBitSizeOf(size_t numElements)
    {
        AutoArrayRecursion autoArrayRecursion;
        fillAutoArrayRecursion(autoArrayRecursion, numElements);

        const size_t bitPosition = 2;
        const size_t autoArrayBitSize = getAutoArrayRecursionBitSizeOf(numElements);
        ASSERT_EQ(autoArrayBitSize, autoArrayRecursion.bitSizeOf(bitPosition));
    }

    void checkInitializeOffsets(size_t numElements)
    {
        AutoArrayRecursion autoArrayRecursion;
        fillAutoArrayRecursion(autoArrayRecursion, numElements);

        const size_t bitPosition = 2;
        const size_t expectedEndBitPosition = bitPosition + getAutoArrayRecursionBitSizeOf(numElements);
        ASSERT_EQ(expectedEndBitPosition, autoArrayRecursion.initializeOffsets(bitPosition));
    }

    void checkReadConstructor(size_t numElements)
    {
        zserio::BitStreamWriter writer(bitBuffer);
        writeAutoArrayRecursionToByteArray(writer, numElements);

        zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
        AutoArrayRecursion autoArrayRecursion(reader);
        checkAutoArrayRecursion(autoArrayRecursion, numElements);
    }

    void checkWriteRead(size_t numElements)
    {
        AutoArrayRecursion autoArrayRecursion;
        fillAutoArrayRecursion(autoArrayRecursion, numElements);

        zserio::BitStreamWriter writer(bitBuffer);
        autoArrayRecursion.write(writer);

        ASSERT_EQ(autoArrayRecursion.bitSizeOf(), writer.getBitPosition());
        ASSERT_EQ(autoArrayRecursion.initializeOffsets(), writer.getBitPosition());

        zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
        AutoArrayRecursion readAutoArrayRecursion(reader);
        checkAutoArrayRecursion(readAutoArrayRecursion, numElements);
    }

    void checkWriteReadFile(size_t numElements)
    {
        AutoArrayRecursion autoArrayRecursion;
        fillAutoArrayRecursion(autoArrayRecursion, numElements);

        const std::string fileName = BLOB_NAME_BASE + std::to_string(numElements) + ".blob";
        zserio::serializeToFile(autoArrayRecursion, fileName);

        AutoArrayRecursion readAutoArrayRecursion = zserio::deserializeFromFile<AutoArrayRecursion>(fileName);
        checkAutoArrayRecursion(readAutoArrayRecursion, numElements);
    }

    static const std::string BLOB_NAME_BASE;
    static const size_t AUTO_ARRAY_LENGTH1;
    static const size_t AUTO_ARRAY_LENGTH2;
    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
};

const std::string AutoArrayStructRecursionTest::BLOB_NAME_BASE =
        "language/array_types/auto_array_struct_recursion_";
const size_t AutoArrayStructRecursionTest::AUTO_ARRAY_LENGTH1 = 5;
const size_t AutoArrayStructRecursionTest::AUTO_ARRAY_LENGTH2 = 10;

TEST_F(AutoArrayStructRecursionTest, bitSizeOfLength1)
{
    checkBitSizeOf(AUTO_ARRAY_LENGTH1);
}

TEST_F(AutoArrayStructRecursionTest, bitSizeOfLength2)
{
    checkBitSizeOf(AUTO_ARRAY_LENGTH2);
}

TEST_F(AutoArrayStructRecursionTest, initializeOffsetsLength1)
{
    checkInitializeOffsets(AUTO_ARRAY_LENGTH1);
}

TEST_F(AutoArrayStructRecursionTest, initializeOffsetsLength2)
{
    checkInitializeOffsets(AUTO_ARRAY_LENGTH2);
}

TEST_F(AutoArrayStructRecursionTest, readConstructorLength1)
{
    checkReadConstructor(AUTO_ARRAY_LENGTH1);
}

TEST_F(AutoArrayStructRecursionTest, readConstructorLength2)
{
    checkReadConstructor(AUTO_ARRAY_LENGTH2);
}

TEST_F(AutoArrayStructRecursionTest, writeReadLength1)
{
    checkWriteRead(AUTO_ARRAY_LENGTH1);
}

TEST_F(AutoArrayStructRecursionTest, writeReadLength2)
{
    checkWriteRead(AUTO_ARRAY_LENGTH2);
}

TEST_F(AutoArrayStructRecursionTest, writeReadFileLength1)
{
    checkWriteReadFile(AUTO_ARRAY_LENGTH1);
}

TEST_F(AutoArrayStructRecursionTest, writeReadFileLength2)
{
    checkWriteReadFile(AUTO_ARRAY_LENGTH2);
}

} // namespace auto_array_struct_recursion
} // namespace array_types

File: test/language/array_types/python/AutoArrayStructRecursionTest.py
import os
import zserio

import ArrayTypes

from testutils import getApiDir

class AutoArrayStructRecursionTest(ArrayTypes.TestCase):
    def testBitSizeOfLength1(self):
        self._checkBitSizeOf(self.AUTO_ARRAY_LENGTH1)

    def testBitSizeOfLength2(self):
        self._checkBitSizeOf(self.AUTO_ARRAY_LENGTH2)

    def testInitializeOffsetsLength1(self):
        self._checkInitializeOffsets(self.AUTO_ARRAY_LENGTH1)

    def testInitializeOffsetsLength2(self):
        self._checkInitializeOffsets(self.AUTO_ARRAY_LENGTH2)

    def testReadLength1(self):
        self._checkRead(self.AUTO_ARRAY_LENGTH1)

    def testReadLength2(self):
        self._checkRead(self.AUTO_ARRAY_LENGTH2)

    def testWriteReadLength1(self):
        self._checkWriteRead(self.AUTO_ARRAY_LENGTH1)

    def testWriteReadLength2(self):
        self._checkWriteRead(self.AUTO_ARRAY_LENGTH2)

    def testWriteReadFileLength1(self):
        self._checkWriteReadFile(self.AUTO_ARRAY_LENGTH1)

    def testWriteReadFileLength2(self):
        self._checkWriteReadFile(self.AUTO_ARRAY_LENGTH2)

    def _checkBitSizeOf(self, numElements):
        autoArrayRecursion = self._createAutoArrayRecursion(numElements)
        bitPosition = 2
        autoArrayRecursionBitSize = 8 + 8 + numElements * (8 + 8)
        self.assertEqual(autoArrayRecursionBitSize, autoArrayRecursion.bitsizeof(bitPosition))

    def _checkInitializeOffsets(self, numElements):
        autoArrayRecursion = self._createAutoArrayRecursion(numElements)
        bitPosition = 2
        expectedEndBitPosition = bitPosition + 8 + 8 + numElements * (8 + 8)
        self.assertEqual(expectedEndBitPosition, autoArrayRecursion.initialize_offsets(bitPosition))

    def _checkRead(self, numElements):
        writer = zserio.BitStreamWriter()
        AutoArrayStructRecursionTest._writeAutoArrayRecursionToStream(writer, numElements)

        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        autoArrayRecursion = self.api.AutoArrayRecursion.from_reader(reader)
        self._checkAutoArrayRecursion(autoArrayRecursion, numElements)

    def _checkWriteRead(self, numElements):
        autoArrayRecursion = self._createAutoArrayRecursion(numElements)
        bitBuffer = zserio.serialize(autoArrayRecursion)

        self.assertEqual(autoArrayRecursion.bitsizeof(), bitBuffer.bitsize)
        self.assertEqual(autoArrayRecursion.initialize_offsets(), bitBuffer.bitsize)

        readAutoArrayRecursion = zserio.deserialize(self.api.AutoArrayRecursion, bitBuffer)
        self._checkAutoArrayRecursion(readAutoArrayRecursion, numElements)

    def _checkWriteReadFile(self, numElements):
        autoArrayRecursion = self._createAutoArrayRecursion(numElements)

        filename = self.BLOB_NAME_BASE + str(numElements) + ".blob"
        zserio.serialize_to_file(autoArrayRecursion, filename)

        readAutoArrayRecursion = zserio.deserialize_from_file(self.api.AutoArrayRecursion, filename)
        self._checkAutoArrayRecursion(readAutoArrayRecursion, numElements)

    def _createAutoArrayRecursion(self, numElements):
        autoArray = []
        for i in range(1, numElements + 1):
            element = self.api.AutoArrayRecursion(i, [])
            autoArray.append(element)

        return self.api.AutoArrayRecursion(0, autoArray)

    def _checkAutoArrayRecursion(self, autoArrayRecursion, numElements):
        self.assertEqual(0, autoArrayRecursion.id)
        autoArray = autoArrayRecursion.auto_array_recursion
        self.assertEqual(numElements, len(autoArray))
        for i in range(1, numElements + 1):
            element = autoArray[i - 1]
            self.assertEqual(i, element.id)
            self.assertEqual(0, len(element.auto_array_recursion))

    @staticmethod
    def _writeAutoArrayRecursionToStream(writer, numElements):
        writer.write_bits(0, 8)
        writer.write_varsize(numElements)
        for i in range(1, numElements + 1):
            writer.write_bits(i, 8)
            writer.write_varsize(0)

    BLOB_NAME_BASE = os.path.join(getApiDir(os.path.dirname(__file__)), "auto_array_struct_recursion_")
    AUTO_ARRAY_LENGTH1 = 5
    AUTO_ARRAY_LENGTH2 = 10
